src/borg/archive.py

```diff
@@ -700,22 +713,39 @@ def process_symlink(self, path, st):
         self.add_item(item)
         return 's'  # symlink
 
+    def chunk_file(self, item, cache, stats, fd, fh=-1, **chunk_kw):
+        checkpoint_number = 1
+        item.chunks = []
+        for data in backup_io_iter(self.chunker.chunkify(fd, fh)):
+            item.chunks.append(cache.add_chunk(self.key.id_hash(data), Chunk(data, **chunk_kw), stats))
+            if self.show_progress:
+                self.stats.show_progress(item=item, dt=0.2)
+            if self.checkpoint_interval and time.time() - self.last_checkpoint > self.checkpoint_interval:
+                checkpoint_item = Item(internal_dict=item.as_dict())
+                checkpoint_item.path += '.checkpoint_%d' % checkpoint_number
+                checkpoint_item.checkpoint = checkpoint_number
+                checkpoint_number += 1
+                self.add_item(checkpoint_item, show_progress=False)
+                self.write_checkpoint()
+                self.last_checkpoint = time.time()
+                # we have saved the checkpoint file, but we will reference the same
+                # chunks also from the checkpoint or the "real" file we save next
+                for chunk in checkpoint_item.chunks:
+                    cache.chunk_incref(chunk.id, stats)
```

Comment:
I think it would be correct if there are no additional increments:

chunks a, b, c, ...

abc
refcounts: 1, 1, 1

checkpoint_1 archive committed

Ok, there is only checkpoint_1 archive pointing to these chunks.

def
refcounts: 1, 1, 1, 1, 1, 1

checkpoint_2 archive committed, checkpoint_1 gone.

Ok, there is only checkpoint_2 archive pointing to these chunks.

Done, commit main archive:
removes checkpoint_2

refcounts 6× (1,)

Ok, only the main archive points there.

---

abc
refcounts: 1, 1, 1

checkpoint_1 archive committed

Ok, there is only checkpoint_1 archive pointing to these chunks.

We crash now.

User restarts backup. File is unchanged.

abc
refcounts: 1+1, 1+1, 1+1

othermain.checkpoint_1 archive commited

Ok: the checkpoint_1 archive of the previous run still exists and uses these chunks as well as our current checkpoint_1

def
refcounts: 1+1, 1+1, 1+1, 1, 1, 1

checkpoint_2 archive committed, checkpoint_1 gone.

\<same\>

Done, commit main archive:
removes checkpoint_2

refcounts 2, 2, 2, 1, 1, 1

Ok, main archive points there, and there is still the checkpoint_1 archive of the old, aborted run around.